nep-16.mediawiki

```diff
@@ -0,0 +1,72 @@
+<pre>
+  NEP: 16
+  Title: NEO Executable Format (NEF)
+  Author: Fernando Díaz Toledano <shargon@gmail.com>, Erik Zhang <erik@neo.org>, Igor Machado Coelho <igormachado@gmail.com>, Li Jianying <lightsever@hotmail.com>, Vitor Nazário Coelho <vncoelho@gmail.com>
+  Type: Standard
+  Status: Accepted
+  Created: 2020-03-31
+</pre>
+
+==Abstract==
+
+A <code>NefFile</code> is a file meeting the standard format for a valid NEO executable file. It provides the basic information about the script, including the compiler and version.
+
+This NEP describes the NEF standard for NEO smart contracts.
+
+==Motivation==
+
+Although we use neo-vm as the virtual machine currently, we should have other scripts in the future. A standard definition for the executable file format can guarantee both the feasibility of other vm code and the validity of the script file.
+
+==Specification==
+
+In the field definitions below, we provide both the definitions of the fields as well as the corresponding functions.
+
+==NefFile Structure==
+
+A <code>nef</code> file consists of a single <code>NefFile</code> structure:
+
+{| class="wikitable sortable" style="width: auto; text-align: center; font-size: smaller; table-layout: fixed;"
+!<nowiki>#</nowiki>
+!size
+!item
+|-
+| 1
+| 4 bytes
+| magic
+|-
+| 2
+| 32 bytes
+| compile
+|-
+| 3
+| 16 bytes
+| version
+|-
+| 4
+| var bytes
+| abi
+|-
+| 5
+| var bytes
+| script
+|}
+
+===magic===
+
+The <code>magic</code> item supplies the magic number identifying the <code>NEF3</code> file format; it has the value <code>0x3346454E</code>.
+
+===compiler===
+
+The <code>compile</code> is a UTF-8 encoded <code>string</code> representing the human readable name of the compiler used to create the data in the <code>script</code> item. Its length is limited to 32 bytes. The name is right side padded by zero's if shorter than 32 bytes.
+
+===version===
+
+The value of <code>version</code> item is determined by a major version number M, a minor version number m, a build number x and a revision number y, each of which is a signed four-byte quantity. Thus we can denote the version of its <code>NefFile</code> as M.m.x.y.
```

Comment:
While we're here, can we make `version` be a string? Different projects can use different versioning schemes and I don't think we can force everyone to four numbers here.